src/Fantomas.Core/SyntaxOak.fs

```diff
@@ -766,49 +766,64 @@ type RecordFieldNode(fieldName: IdentListNode, equals: SingleTextNode, expr: Exp
     member val Equals = equals
     member val Expr = expr
 
-type ExprRecordNode
+[<AbstractClass>]
+type ExprRecordNode(openingBrace: SingleTextNode, fields: RecordFieldNode list, closingBrace: SingleTextNode, range) =
+    inherit NodeBase(range)
+
+    member val OpeningBrace = openingBrace
+    member val Fields = fields
+    member val ClosingBrace = closingBrace
+    member x.HasFields = List.isNotEmpty x.Fields
+
+type ExprCopyableRecordNode
     (
         openingBrace: SingleTextNode,
-        extra: RecordNodeExtra,
+        copyInfo: Expr option,
         fields: RecordFieldNode list,
         closingBrace: SingleTextNode,
         range
     ) =
-    inherit NodeBase(range)
+    inherit ExprRecordNode(openingBrace, fields, closingBrace, range)
+
+    member val CopyInfo = copyInfo
 
     override val Children: Node array =
         [| yield openingBrace
-           yield! noa (RecordNodeExtra.Node extra)
+           yield! copyInfo |> Option.map Expr.Node |> noa
            yield! nodes fields
            yield closingBrace |]
 
-    member val OpeningBrace = openingBrace
-    member val Extra = extra
-    member val Fields = fields
-    member val ClosingBrace = closingBrace
+    member x.HasFields = List.isNotEmpty x.Fields
 
-type ExprAnonRecordNode
+type ExprInheritRecordNode
     (
-        isStruct: bool,
         openingBrace: SingleTextNode,
-        copyInfo: Expr option,
+        inheritConstructor: InheritConstructor,
         fields: RecordFieldNode list,
         closingBrace: SingleTextNode,
         range
     ) =
-    inherit NodeBase(range)
+    inherit ExprRecordNode(openingBrace, fields, closingBrace, range)
+
+    member val InheritConstructor = inheritConstructor
 
     override val Children: Node array =
         [| yield openingBrace
-           yield! noa (Option.map Expr.Node copyInfo)
+           yield InheritConstructor.Node inheritConstructor
            yield! nodes fields
            yield closingBrace |]
 
+type ExprAnonRecordNode
+    (
+        isStruct: bool,
```

Comment:
You don't really need the boolean if you were to rename this to `ExprAnonRecordStructNode`.
A non-struct anonymous record would fit perfectly in `ExprCopyableRecordNode`, so you only really need a new type for when it has the `struct` in it.

Reply:
There are still a few differences in non-struct records that I've noticed, see `genMultilineAnonRecordCrampedFields` (which I'll also inline like the other one-offs you mentioned). If those can be consolidated, that would be great, but I wasn't able to get everything to work without those changes